dbms/src/Flash/Disaggregated/RNPagePreparer.cpp

```diff
@@ -0,0 +1,197 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Common/Exception.h>
+#include <Common/TiFlashMetrics.h>
+#include <Flash/Coprocessor/CHBlockChunkCodec.h>
+#include <Flash/Disaggregated/RNPagePreparer.h>
+#include <Flash/Disaggregated/RNPageReceiver.h>
+#include <IO/IOThreadPool.h>
+#include <Storages/DeltaMerge/DeltaMergeHelpers.h>
+#include <Storages/DeltaMerge/Remote/RNRemoteReadTask.h>
+#include <common/logger_useful.h>
+
+#include <future>
+
+namespace DB
+{
+
+RNPagePreparer::RNPagePreparer(
+    DM::RNRemoteReadTaskPtr remote_read_tasks_,
+    std::shared_ptr<RNPageReceiver> receiver_,
+    const DM::ColumnDefinesPtr & columns_to_read,
+    size_t max_streams_,
+    const String & req_id,
+    const String & executor_id,
+    bool do_prepare_)
+    : threads_num(max_streams_)
+    , do_prepare(do_prepare_)
+    , remote_read_tasks(std::move(remote_read_tasks_))
+    , receiver(std::move(receiver_))
+    , live_persisters(threads_num)
+    , state(PageReceiverState::NORMAL)
+    , total_rows(0)
+    , total_pages(0)
+    , exc_log(Logger::get(req_id, executor_id))
+{
+    assert(columns_to_read != nullptr);
+    decoder_ptr = std::make_unique<CHBlockChunkCodec>(DM::toEmptyBlock(*columns_to_read));
+
+    try
+    {
+        for (size_t index = 0; index < threads_num; ++index)
+        {
+            auto task = std::make_shared<std::packaged_task<void()>>([this, index] {
+                persistLoop(index);
+            });
+            persist_threads.emplace_back(task->get_future());
+
+            IOThreadPool::get().scheduleOrThrowOnError([task] { (*task)(); });
+        }
+    }
+    catch (...)
+    {
+        tryLogCurrentException(exc_log, __PRETTY_FUNCTION__);
+        throw;
+    }
+}
+
+RNPagePreparer::~RNPagePreparer() noexcept
+{
+    for (auto & task : persist_threads)
+    {
+        try
+        {
+            task.get();
+        }
+        catch (...)
+        {
+            tryLogCurrentException(exc_log, __PRETTY_FUNCTION__);
+        }
+    }
+}
+
+void RNPagePreparer::persistLoop(size_t idx)
+{
+    LoggerPtr log = exc_log->getChild(fmt::format("persist{}", idx));
+
+
+    bool meet_error = false;
+    String local_err_msg;
+
+    while (!meet_error)
+    {
+        try
+        {
+            // no more results
+            if (!consumeOneResult(log))
+                break;
+        }
+        catch (...)
+        {
+            meet_error = true;
+            local_err_msg = getCurrentExceptionMessage(false);
+        }
+    }
+
+    downloadDone(meet_error, local_err_msg, log);
+
+    // try to do some preparation for speed up reading
```

Comment:
Personally I'm not a fan of conditional preparing. It introduce conditions to our flows (some task's delta index may be prepared and some may be not), make it harder to troubleshoot and harder to understand. It seems fine if we just prepare the delta index for all tasks before reading, which will not introduce extra latencies.

Anyway we can simplify later.